rtengine/lensmetadata.cc

```diff
@@ -0,0 +1,1035 @@
+    void parse()
+    {
+        has_dist = has_ca = has_vign = false;
+
+        auto &exif = metadata.exifData();
+
+        auto it = exif.findKey(Exiv2::ExifKey("Exif.SubImage1.OpcodeList3"));
+
+        if (it != exif.end()) {
+            std::vector<Exiv2::byte> buf;
+            buf.resize(it->value().size());
+            it->value().copy(&buf[0], Exiv2::invalidByteOrder);
+
+            const Exiv2::byte *data = &buf[0];
+            uint32_t num_entries = Exiv2::getULong(data, Exiv2::bigEndian);
+            size_t idx = 4;
+
+            for (size_t i = 0; i < num_entries && idx < buf.size(); ++i) {
+                uint32_t opcodeID = Exiv2::getULong(data + idx, Exiv2::bigEndian);
+                idx += 4;
+                idx += 4; // version
+                uint32_t flags = Exiv2::getULong(data + idx, Exiv2::bigEndian);
+                idx += 4;
+                size_t paramSize = Exiv2::getULong(data + idx, Exiv2::bigEndian);
+                idx += 4;
+
+                if (idx + paramSize > buf.size()) {
+                    throw std::runtime_error("error parsing DNG OpcodeList3");
+                }
+
+                if (processed_opcodes.find(opcodeID) != processed_opcodes.end()) {
+                    // we currently handle only one opcode per type and ignore next ones if provided.
+                    if (settings->verbose) {
+                        std::printf("DNG OpcodeList3 %s opcode %d already processed\n", flags & 1 ? "optional" : "mandatory", opcodeID);
+                    }
+
+                    idx += paramSize;
+                    continue;
+                }
+
+                processed_opcodes.insert(opcodeID);
+
+                // we currently handle only one dist correction
+                if (opcodeID == 1 && !has_dist) { // WarpRectilinear
+
+                    planes = Exiv2::getULong(data + idx, Exiv2::bigEndian);
+
+                    if ((planes != 1) && (planes != 3)) {
+                        throw std::runtime_error("cannot parse DNG WarpRectilinear");
+                    }
+
+                    for (int p = 0; p < planes; p++) {
+                        for (int i = 0; i < 6; i++) {
+                            warp_rectilinear[p][i] = Exiv2::getDouble(data + idx + 4 + 8 * (i + p * 6), Exiv2::bigEndian);
+                        }
+                    }
+
+                    crx_d = Exiv2::getDouble(data + idx + 4 + 8 * (0 + planes * 6), Exiv2::bigEndian);
+                    cry_d = Exiv2::getDouble(data + idx + 4 + 8 * (1 + planes * 6), Exiv2::bigEndian);
+
+                    has_dist = true;
+                    if (planes == 3) {
+                        has_ca = true;
+                    }
+
+                    // we currently handle only one vignetting correction
+                } else if (opcodeID == 3 && !has_vign) { // FixVignetteRadial
+                    size_t start = idx;
+                    size_t end = idx + 7 * 8;
+                    if (end <= buf.size()) {
```

Comment:
Should this be `>`?

Reply:
It checks that the expected end is inside the buffer size.